great_expectations/core/batch_definition.py

```diff
@@ -26,6 +27,7 @@ class BatchDefinition(pydantic.BaseModel):
     id: Optional[str] = None
     name: str
     partitioner: Optional[Partitioner] = None
+    batching_regex: Optional[re.Pattern] = None
```

Comment:
I'm wondering since we don't use `batching_regex` is some assets if we should really have multiple classes. That is we could have a baseclass or protocol without `batching_regex` and a concrete implementation with and without it. Then we wouldn't need to do the runtime checks for invalid fields for particular assets. This isn't necessary for this PR.

Reply:
I think I lean toward multiple classes as well. But before doing that, I think we need to think through some of the implications. e.g. does an asset have a `batch_definition: BatchDefinitionWithBatchingRegex`, or does it have a `batch_definition: BatchDefinition` and then do runtime checks to ensure it has a batching regex.